socks5.c

```diff
@@ -0,0 +1,337 @@
+static void dill_socks5_hclose(struct dill_hvfs *hvfs);
+static int dill_socks5_bsendl(struct dill_bsock_vfs *bvfs,
+    struct dill_iolist *first, struct dill_iolist *last, int64_t deadline);
+static int dill_socks5_brecvl(struct dill_bsock_vfs *bvfs,
+    struct dill_iolist *first, struct dill_iolist *last, int64_t deadline);
+
+// SOCKS5 as defined in RFC 1928
+
+struct dill_socks5_sock {
+    struct dill_hvfs hvfs;
+    struct dill_bsock_vfs bvfs;
+    /* Underlying SUFFIX socket. */
+    int u;
+    unsigned int mem : 1;
+    unsigned int server : 1;
+    unsigned int connect : 1;
+    //struct dill_suffix_storage suffix_mem;
+    //struct dill_term_storage term_mem;
+    char rxbuf[1024];
+};
+
+DILL_CHECK_STORAGE(dill_socks5_sock, dill_socks5_storage)
+
+static void *dill_socks5_hquery(struct dill_hvfs *hvfs, const void *type) {
+    struct dill_socks5_sock *obj = (struct dill_socks5_sock*)hvfs;
+    if(type == dill_socks5_type) return obj;
+    if(type == dill_bsock_type) return &obj->bvfs;
+    errno = ENOTSUP;
+    return NULL;
+}
+
+static int dill_socks5_clientauth(int s, const char *username, const char *password,
+      int64_t deadline) {
+    // ensure socket is attached as socks5 
+    struct dill_socks5_sock *obj = dill_hquery(s, dill_socks5_type);
+    if(dill_slow(!obj)) return -1;
+    // validate input, specifically user and pass, which may be NULL
+    if (username) {
+        if (strlen(username) > 255) {errno = EINVAL; return -1;}
+    }
+    if (password) {
+        if (strlen(password) > 255) {errno = EINVAL; return -1;}
+    }
+    // version identifier/method selection request
+    // client passes list of desired auth methods
+    uint8_t vims[4];
+    size_t vims_len;
+    vims[0] = 0x05; // VER = 5
+    vims[2] = 0x00; // METHOD = NO AUTH REQ'D
+    vims[3] = 0x02; // METHOD = USERNAME/PASSWORD
+    // TODO: Add support for GSSAPI Auth
+    // if we don't have both username and password only request NO AUTH
+    if ((!username) || (!password)) {
+        vims[1] = 0x01; // NMETHODS = 1 (NO AUTH)
+        vims_len = 3;
+    } else {
+        vims[1] = 0x02; // NMETHODS = 2 (NO AUTH or USER/PASS)
+        vims_len = 4;
+    }
+    //for (int i = 0; i < vims_len; i++) {printf("%02X ", vims[i]);} printf(">\n");
+    int err = dill_bsend(obj->u, (void *)vims, vims_len, deadline);
+    if (dill_slow(err)) {return -1;}
+    // version identifier/method response
+    // server responds with chosen auth method (or error)
+    uint8_t vimr[2];
+    err = dill_brecv(obj->u, (void *)vimr, 2, deadline);
+    if (dill_slow(err)) {return -1;}
+    //for (int i = 0; i < 2; i++) {printf("%02X ", vimr[i]);} printf("<\n");
+    // validate VER in response
+    if (dill_slow(vimr[0] != 0x05)) {errno = EPROTO; return -1;}
+    // if proxy accepted NO AUTH REQ'D, we're done
+    if (vimr[1] == 0x00) {return 0;}
+    if (vimr[1] == 0x02) {
+        dill_assert(!!username);
```

Comment:
Handle this case decently.

Reply:
replaced with switch() statement, I agree much cleaner